### ISSUE TITLE
":terminal emacs" doesn't quit with ^X^C

### ISSUE THREAD

**Issue by mattn** [vim/vim#1988]
```c
may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);
```

If disable this line, emacs quit correctly.

**Comment by mattn** [comment]
How about to check ISIG ?

```diff
diff --git a/src/terminal.c b/src/terminal.c
index 149679e..f6c4dee 100644
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -66,6 +66,18 @@
 
 #include "vim.h"
 
+#if defined(HAVE_TERMIOS_H)
+#  include <termios.h>
+#else
+# ifdef HAVE_TERMIO_H
+#  include <termio.h>
+# else
+#  ifdef HAVE_SGTTY_H
+#   include <sgtty.h>
+#  endif
+# endif
+#endif
+
 #if defined(FEAT_TERMINAL) || defined(PROTO)
 
 #ifndef MIN
@@ -1363,7 +1375,25 @@ terminal_loop(void)
 	    continue;
 
 #ifdef UNIX
-	may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);
+	{
+	    int	accept_int = 0;
+# if defined(HAVE_TERMIOS_H)
+	    struct termios told;
+	    tcgetattr(read_cmd_fd, &told);
+	    accept_int = told.c_iflag & ISIG;
+# elif defined(HAVE_TERMIO_H)
+	    struct termio told;
+	    ioctl(read_cmd_fd, TCGETA, &told);
+	    accept_int = told.c_iflag & ISIG;
+# else
+	    struct sgttyb told;
+	    ioctl(read_cmd_fd, TIOCGETP, &told);
+	    accept_int = told.sg_flags & ISIG;
+# endif
+	    if (!accept_int)
+		may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);
+	}
+
 #endif
 #ifdef WIN3264
 	/* On Windows winpty handles CTRL-C, don't send a CTRL_C_EVENT.
```

**Comment by mattn** [comment]
Sorry, revered

	    if (!accept_int)
		may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);

**Comment by mattn** [comment]
Oops, confusing. Maybe variable name is wrong.
```diff
diff --git a/src/terminal.c b/src/terminal.c
index 149679e..88d6796 100644
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -66,6 +66,18 @@
 
 #include "vim.h"
 
+#if defined(HAVE_TERMIOS_H)
+#  include <termios.h>
+#else
+# ifdef HAVE_TERMIO_H
+#  include <termio.h>
+# else
+#  ifdef HAVE_SGTTY_H
+#   include <sgtty.h>
+#  endif
+# endif
+#endif
+
 #if defined(FEAT_TERMINAL) || defined(PROTO)
 
 #ifndef MIN
@@ -1363,7 +1375,25 @@ terminal_loop(void)
 	    continue;
 
 #ifdef UNIX
-	may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);
+	{
+	    int	handle_intr = 0;
+# if defined(HAVE_TERMIOS_H)
+	    struct termios told;
+	    tcgetattr(read_cmd_fd, &told);
+	    handle_intr = told.c_iflag & ISIG;
+# elif defined(HAVE_TERMIO_H)
+	    struct termio told;
+	    ioctl(read_cmd_fd, TCGETA, &told);
+	    handle_intr = told.c_iflag & ISIG;
+# else
+	    struct sgttyb told;
+	    ioctl(read_cmd_fd, TIOCGETP, &told);
+	    handle_intr = told.sg_flags & ISIG;
+# endif
+	    if (handle_intr)
+		may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);
+	}
+
 #endif
 #ifdef WIN3264
 	/* On Windows winpty handles CTRL-C, don't send a CTRL_C_EVENT.
```

**Comment by mattn** [comment]
Hmm, something wrong.

**Comment by brammool** [comment]
Yashiro Matsumoto wrote:

> ```c
> may_send_sigint(c, curbuf->b_term->tl_job->jv_pid, 0);
> ```
> 
> If disable this line, emacs quit correctly.

Of course it is very important so support quitting emacs.

Thinking about this: If the termios indicate a CTRL-C causes an
interrupt, we can just send a CTRL-C.  If the termios that CTRL-C does
not cause an interrupt, then the command must have changed that and
wants to handle CTRL-C by itself.

I tried running "cat", and without may_send_sigint() pressing CTRL-C
does end it.

It's similar to the problem we had for MS-Windows.  I think the best is
to indeed remove may_send_sigint().  We would another key to kill the
job.  I think we can use CTRL-W CTRL-C.

-- 
ARTHUR: What are you going to do. bleed on me?
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
closed by [patch 8.0.0960](https://github.com/vim/vim/commit/8e539c51c31461fc602071373bdff9edd1aed25f)
